### ISSUE TITLE
Refactor builders

### ISSUE THREAD

**Issue by dieppa** [cloudyrock/mongock#36]
Considerable amount of duplicated code in the builders

**Comment by droidkfx** [comment]
Hey, you mentioned this issue in a PR I created. Can you provide some more information on what you want? Is the intention just to minimize the amount of code in the builders and move all duplications to a base class, or utility class? I would be happy to implement that just want to make sure I know what you need. Let me know.

**Comment by dieppa** [comment]
Sorry @droidkfx , I didn't see it.

Ok, so the goal of this is the following:

- As you mention, reduce code duplication
- Improve hierarchy/inheritance, so it's more readable, reusable and extensible 
- Increase encapsulation. By having the builders in the class itself, setters that shouldn't be exposed, can be private 
- Facilitate tests...less and easier
- etc.

Anyway, there is another [issue](https://github.com/cloudyrock/mongock/issues/35) which is related to this one and there could be some conflicts between both.

**Comment by dieppa** [comment]
If you decide to do this, please let me know to discuss the approach, so we have the same idea ;)

**Comment by Danjenator** [comment]
I'm not sure if @droidkfx started on this, but if not I can work on this.

**Comment by dieppa** [comment]
Hi @Danjenator. Feel free to jump on it. Just for you to know, this issue is a bit undefined. It depends on each person's point of view, but as long as your change improves the code and fits best practices, even if it's not everything, or the way I thought, you will be fine ;)

**Comment by Danjenator** [comment]
Thanks, that sounds good. I used this at a client, but couldn't use the setLockQuickConfig() method with Spring boot. Now that I'm on the bench, I thought I'd fix it. I'll have a pr coming shortly, so let me know if you want me to change how I implemented this.

Thank you for forking this from MongoBee and getting it up to date! I was pulling my hair out for a few days trying to get that to work before I realized it was no longer being maintained.

**Comment by dieppa** [comment]
Yeah, you are right, it returns the parent class. You can either (safely) cast it as temporally work aroung, as you know it returns a SpringBootMongock, or you can actually fix it.

I am creating an issue specially for this, please feel free to jump on it, but, please, let me know and I assign the issue to you.

**Comment by Danjenator** [comment]
I addressed the issue in the PR for this feature. Let me know if there's anything else I need do for this.

**Comment by dieppa** [comment]
oh sorry, didn't see the PR, it just came through now. If you can amend it and point to [this new issue](https://github.com/cloudyrock/mongock/issues/59) I just created for this, it would be great.

I can expect to have it reviewed by Saturday ;)

**Comment by dieppa** [comment]
@Danjenator , I just had a very quickly look to your PR. The intention is good, but I don't agree with the solution. What about simulating the _self type_  mechanism of some languages like Perl, by making a generic class/interface like:

```
public [abstract class | interface] MongockBuilder<RETURN_TYPE extends MongockBuilder> {

 public RETURN_TYPE setLockQuickConfig() {...}
...
}
```

```
public class SpringBootMongockBuilder extends   MongockBuilder<SpringBootMongockBuilder> {
...
}
```


That is much more simpler and quicker. You cannot do that directly as MongockBuilder is not abstract, but that has easy fix, I leave it on you.


Please, avoid using .Net sintaxis like "I" prefix for interface ;D

**Comment by dieppa** [comment]
@Danjenator ,in this [branch](https://github.com/cloudyrock/mongock/tree/feature/issue_59) you can see what I mean. I did it quickly, so there may be some improvements, so feel free to take it as base line for your PR and extend it with the improvements you believe are needed. But let me know, otherwise I refine that branch and merge it.

**Comment by Danjenator** [comment]
@dieppa Sorry for dropping off the face of the earth for a couple days. I see you merged your pr and closed the issues. I'm assuming all is well then. Sorry to only highlight an issue and not solve it.

**Comment by dieppa** [comment]
No problem. Yeah, it's fixed and closed. I am still pending to release a new version, which will be done soon, hopefully this evening
